vm/chef/cookbooks/postgresql/files/pgcluster-utils

```diff
@@ -0,0 +1,190 @@
+#######################################
+function setup_bucardo() {
+  which bucardo > /dev/null
+  if [[ "$?" -ne 0 ]]; then
+    echo "Bucardo not found."
+    exit 1
+  fi
+
+  mkdir -p /var/run/bucardo
+}
+
+#######################################
+# Executes a given command locally using default postgres user.
+# Arguments:
+#   Command to be ran.
+#######################################
+function execute_query() {
+  local -r command="$1"
+
+  if [[ -z "${PGPASSWORD}" ]]; then
+    echo "PGPASSWORD environment variable is not set."
+    exit 1
+  fi
+
+  echo "Executing query: ${command}"
+  psql "postgresql://${postgres_username}:${PGPASSWORD}@localhost" -c "${command}"
+}
+
+#######################################
+# Ensure all given instances have the replication table and bucardo user can access it.
+# All replicas will be tested, but not master one.
+# Arguments:
+#   Database name.
+#   Instances list.
+#######################################
+function await_for_replicas() {
+  local -r database_name="$1"
+  local -r instances="$2"
+  local connected=""
+  local instance_id=""
+
+  for instance in $instances; do
+    instance_id="$(echo ${instance} | awk -F '-' '{ print $NF }')"
+    connected="false"
+
+    # Skip for master node
+    if [[ "${instance_id}" == "0" ]]; then
+      continue
+    fi
+
+    echo "Trying to connect to instance ${instance}..."
+    while [[ "${connected}" == "false" ]]; do
+      # Avoid to fail script as we are checking exit-code.
+      set +e
+      su - postgres -c "psql -U bucardo -d ${database_name} -h \"${instance}\" -c \"SELECT 1 FROM ${sample_table};\"" \
+        >/dev/null 2>&1
+      set -e
+
+      if [[ "$?" -eq 0 ]]; then
+        echo "Connected to the instance: ${instance}."
+        connected="true"
+      else
+        echo "Connection to ${instance} failed."
+      fi
+    done
+  done
+}
+
+#######################################
+# Enroll all the instances databases to the replication service.
+# Arguments:
+#   Database name.
+#   Instances list.
+#######################################
+function add_databases_to_replication() {
+  local -r database_name="$1"
+  local -r instances="$2"
+  local instance_id=""
+  local node_name=""
+
+  for instance in $instances; do
+    # Node name should be short and without special chars.
+    instance_id="$(echo ${instance} | awk -F '-' '{ print $NF }')"
+    node_name="node${instance_id}"
+
+    echo "Adding ${instance} to bucardo..."
+
+    # Add database server
+    bucardo add database "${node_name}" \
+      dbname="${database_name}" \
+      host="${instance}"
+
+    # Add all available tables
+    bucardo add all table % db="${node_name}"
+  done
+}
+
+#######################################
+# Add the replication syncs permutating all possible combinations.
+# Arguments:
+#   Instances list.
+#######################################
+function add_replication_syncs() {
+  local -r instances_list="$1"
+  local source_instance_id=""
```

Comment:
why do you need to declare those vars beforehand?

Reply:
Done.